Use proper repositories for centos/rhel. Use gpg key. Use mainline repo if desired

wernerb opened the pull request.

geerlingguy commented:
Ah, noticed you're setting up centos vs redhat in here too—instead of having two separate template files and separate tasks to add them, you can do one template/task and use a variable to switch the repo URL between `rhel` and `centos`. I'll leave a few notes in the PR itself.

geerlingguy commented on tasks/setup-RedHat.yml:
```diff
@@ -1,11 +1,36 @@
+  rpm_key: state=present key="{{ nginx_el_repo_gpg_key | upper }}"
+  when: gpg_key_download.changed
+
+- name: (RedHat) Enable nginx repo.
```
No need for two tasks to do this—you can do it all in one, without a `when`... see my comments in the template file for the rest of the change required.

geerlingguy commented on templates/nginx.centos.repo.j2:
```diff
@@ -0,0 +1,13 @@
+{% if nginx_el_install_mainline %}
+[nginx]
+name=nginx repo mainline
+baseurl=http://nginx.org/packages/mainline/centos/$releasever/$basearch/
```
You can change the `centos` here to: `{{ ansible_distribution | lower }}`

I don't have a RHEL server around to verify it outputs "RHEL" though... can someone else check?

Worst case you could add a little conditional logic to print `centos` or `rhel`. I'd rather have one file.

geerlingguy commented on README.md:
```diff
@@ -81,6 +81,10 @@ Set as the `proxy_cache_path` directive in the `nginx.conf` file. By default, th
 
 (For Debian/Ubuntu only) Allows you to set a different repository for the installation of Nginx. As an example, if you are running Debian's wheezy release, and want to get a newer version of Nginx, you can install the `wheezy-backports` repository and set that value here, and Ansible will use that as the `-t` option while installing Nginx.
 
+    nginx_el_install_mainline: yes
```
Can you change this variable (just to make it more intuitive) to `nginx_install_redhat_branch` with values of `mainline` or `stable` (and then update the logic in the template file accordingly)? I'd also like the default to be `stable`, since I prefer stability over the latest in this role.

wernerb commented:
Hi! The default was already set to stable. I changed the variable name according to your suggestion.
Also changed the two template files to a single one. My personal style is too never really use template logic too much :)

wernerb commented:
About the gpg stuff. This is the reasoning:
The three steps do:

1. Check if key id (fingerprint taken from nginx_signed_key) present in rpm. If not found then we mark it as changed and only then go to the next step. If key is present we skip the next two steps.
2. At `+  rpm_key: state=present key=http://nginx.org/keys/nginx_signing.key` it becomes interesting.. Here nginx sign key is downloaded from http and not https. A MITM could override nginx and download the wrong key at this point.
3. Because there's a risk the wrong key is installed we now have to check if the key with out fingerprint key id is present after the download. This last step is usually used in ansible to remove a key. In this case we use it to check for presence. If however, it is not present (which would be the case in a MITM) it throws an error saying there's no file that can be found to set the key.

It would have been simpler if nginx had used SSL and if ansible had a separate gpg module, but this works.
Because of the when: x changed we avoid having to download the (potentially wrong) nginx signing key in step 2.

wernerb commented:
Hi @geerlingguy , i was wondering if you could perhaps review the changes?

wernerb commented:
* Bump :)

unn commented:
Might want to add an selinux policy, since it will crush souls otherwise.

geerlingguy commented:
@unn - Thanks for the note. @wernerb - can you address that question?

LukeHandle commented:
* `{{ (ansible_distribution == 'CentOS') | ternary('centos', 'rhel') }}` is a lot cleaner (if matches, choose "centos", otherwise use "rhel").
 * https://nginx.org/keys/nginx_signing.key is _now_ available via HTTPS - any reason not to just link from within repo file (`gpgkey`)/skip the "manual" verification
 * @unn can you further clarify this (sorry), do you mean that we should set the `seuser` attribute when creating the repo (eg. to "system_u")?

unn commented:
@geerlingguy don't exactly recall the scope of my request originally. I would be willing to bet that I was looking for a specific selinux policy for nginx shipped in the role itself.

cornfeedhobo commented:
AFAIK, the policy would have to be specific to things that would be set in the config, e.g. access to a directory, writing to a php-fpm socket, etc.

unn commented:
@cornfeedhobo and that's the other problem... maybe something along the lines of a message like, 

> hey we noticed you're running selinux, you probably need to put a policy in place for nginx since we can't automatically generate one.
